_posts/2018-01-03-Gauge-automated-testing.md

```diff
@@ -0,0 +1,286 @@
+        this.passwordField = form.findElement(By.name("user[password]"));
+        this.passwordConfirmField = form.findElement(By.name("user[password_confirmation]"));
+        this.commitButton = form.findElement(By.name("commit"));
+    }
+}
+```
+
+Step definition:
+```java
+public class CustomerSignup {
+
+    private SignUpPage signUpPage = new SignUpPage();
+
+    @Step("Sign up as <customer> with email <customer@example.com> and <password>")
+    public void registerCustomerWith(String customer, String email, String password) {
+        this.signUpPage.usernameField.sendKeys(customer);
+        this.signUpPage.emailField.sendKeys(email);
+        this.signUpPage.passwordField.sendKeys(password);
+        this.signUpPage.passwordConfirmField.sendKeys(password);
+        this.signUpPage.commitButton.click();
+    }
+}
+```
+
+The great benefit of this approach is that you can reuse a lot of the code.
+You only have to locate the elements once instead of in every step.
+Functions can be added to the page objects as well.
+Suppose you have a dropdown, you first have to click to open it and then select an option from the list.
+You can write a function doing all that.
+In your step definition, you then simply have to call that function to select something from a dropdown.
+It's a good way to avoid too much code duplication.
+
+### Concepts
+If you find yourself repeating the same sequence of steps over and over,
+you could combine those steps into one step using concepts.
+These are also written in MarkDown and you can pass arguments the same way as in the Java `@Step` annotations.
+They should be placed in the `/specs/concepts` folder and use the *.cpt extension.
+
+```markdown
+# Sign up a new customer with name <name> email <email> and <password>
+* Sign up as <name> with email <email> and <password>
+* Show a message "Thank you for signing up! You are now logged in."
+```
+
+Cucumber only offers this feature in some implementations, they don't in Cucumber.js for example.
+They refuse to implement it because they believe this creates too much abstraction and makes you lose sight of the overall picture.
+Gauge does offer it, so it's up to you whether you want to make use of it or not.
+
+# Running the specs
+Since the project is setup with Maven, the tests can be run with `mvn test`.
+However, if you want to pass any arguments, you'll need to use `mvn gauge:execute` instead.
+
+### Tags
+
+You may have noticed in the spec files that tags can be added.
+They can be used to run only certain specs.
+
+`mvn gauge:execute -DspecsDir=specs -Dtags="sign-up & customer"`
+
+### Parallel
+So far, Gauge didn't stand out from other automated testing solutions.
+Although there's one thing that really does stand out and that's how easy it is to use parallel execution!
+
+`mvn gauge:execute -DspecsDir=specs -DinParallel=true`
+
+Running this command will look how many CPU cores your computer has and start up that many streams.
```

Comment:
I think it sounds a bit better like this:
Running this command will start up a stream for each CPU core your computer has.